File: push-openapi-net-demo/Custom_Signature/Sign.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Security.Cryptography;

namespace AliCloudPush
{
    class Sign
    {
        private const String SEPARATOR = "&";
        public static string sign(string method, SortedDictionary<string, object> parameters,string secret)
        {
            StringBuilder canonicalizedQueryString = new StringBuilder();
            foreach (var p in parameters)
            {
                canonicalizedQueryString.Append("&")
                .Append(PopEncode.popEncode(p.Key)).Append("=")
                .Append(PopEncode.popEncode(p.Value.ToString()));
            }
            StringBuilder stringToSign = new StringBuilder();
            stringToSign.Append(method.ToString());
            stringToSign.Append(SEPARATOR);
            stringToSign.Append(PopEncode.popEncode("/"));
            stringToSign.Append(SEPARATOR);
            stringToSign.Append(PopEncode.popEncode(
                    canonicalizedQueryString.ToString().Substring(1)));

            var algorithm = KeyedHashAlgorithm.Create("HMACSHA1");
            algorithm.Key = Encoding.UTF8.GetBytes(secret.ToCharArray());
            return Convert.ToBase64String(algorithm.ComputeHash(Encoding.UTF8.GetBytes(stringToSign.ToString().ToCharArray())));
        }
    }
}

File: OpenApi1.0/push-openapi-php-demo/Custom_Signature/sign.php
<?php
	include 'pop_encode.php';
	function sign($method,$params,$secret)
	{
		ksort($params);
		$canonicalizedQueryString = '';
		foreach($params as $key => $value)
            	{
                        $canonicalizedQueryString .= '&' . pop_encode($key). '=' . pop_encode($value);
            	}
		$stringToSign = $method . '&%2F&' . pop_encode(substr($canonicalizedQueryString, 1));
		return base64_encode(hash_hmac('sha1', $stringToSign, $secret, true));
	}
?>
